nexus/src/app/quiesce.rs

```diff
@@ -0,0 +1,369 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Manage Nexus quiesce state
+
+use super::saga::SagaExecutor;
+use assert_matches::assert_matches;
+use chrono::Utc;
+use nexus_db_queries::authz;
+use nexus_db_queries::context::OpContext;
+use nexus_db_queries::db::DataStore;
+use nexus_types::internal_api::views::QuiesceState;
+use nexus_types::internal_api::views::QuiesceStatus;
+use omicron_common::api::external::LookupResult;
+use omicron_common::api::external::UpdateResult;
+use std::sync::Arc;
+use std::time::Instant;
+use tokio::sync::watch;
+
+impl super::Nexus {
+    pub async fn quiesce_start(&self, opctx: &OpContext) -> UpdateResult<()> {
+        opctx.authorize(authz::Action::Modify, &authz::QUIESCE_STATE).await?;
+        let started = self.quiesce.send_if_modified(|q| {
+            if let QuiesceState::Running = q {
+                let time_requested = Utc::now();
+                let time_waiting_for_sagas = Instant::now();
+                *q = QuiesceState::WaitingForSagas {
+                    time_requested,
+                    time_waiting_for_sagas,
+                };
+                true
+            } else {
+                false
+            }
+        });
+        if started {
+            tokio::spawn(do_quiesce(
+                self.quiesce.clone(),
+                self.sagas.clone(),
+                self.datastore().clone(),
+            ));
+        }
+        Ok(())
+    }
+
+    pub async fn quiesce_state(
+        &self,
+        opctx: &OpContext,
+    ) -> LookupResult<QuiesceStatus> {
+        opctx.authorize(authz::Action::Read, &authz::QUIESCE_STATE).await?;
+        let state = self.quiesce.borrow().clone();
+        let sagas_running = self.sagas.sagas_running();
+        let db_claims = self.datastore().claims_held();
```

Comment:
Reading this, and seeing how much state needs to be up-and-running for us to actually do quiesce, I've filed https://github.com/oxidecomputer/omicron/issues/8750.

(This isn't a request for changes here - I think this makes sense - but I think we'll need to re-organize our Nexus construction for quiesce to work with "rebooting, but old, Nexus zones")